doc-content/shared-kie-docs/src/main/asciidoc/KieServer/kie-server-rest-api-solvers-ref.adoc

```diff
@@ -0,0 +1,419 @@
+{
+  "solver-instance": {
+    "container-id": "employee-rostering",
+    "solver-id": "solver1",
+    "solver-config-file": "employeerostering/employeerostering/EmployeeRosteringSolverConfig.solver.xml",
+    "status": "NOT_SOLVING",
+    "score": {
+      "value": null,
+      "scoreClass": null
+    },
+    "best-solution": null
+  }
+}
+----
+
+== [GET] /containers/{containerId}/solvers/{solverId}
+
+Returns the current state of the specified solver within the specified KIE container.
+
+.Request parameters
+[cols="25%,45%,15%,15%", frame="all", options="header"]
+|===
+|Name
+|Description
+|Type
+|Requirement
+
+|`containerId`
+|ID of the KIE container where the solver resides
+|String
+|Required
+
+|`solverId`
+|ID of the solver
+|String
+|Required
+|===
+
+.Example server response (JSON)
+[source,json]
+----
+{
+  "solver-instance": {
+    "container-id": "employee-rostering",
+    "solver-id": "solver1",
+    "solver-config-file": "employeerostering/employeerostering/EmployeeRosteringSolverConfig.solver.xml",
+    "status": "NOT_SOLVING"
+  }
+}
+----
+
+== [DELETE] /containers/{containerId}/solvers/{solverId}
+
+⁠Disposes the specified solver within the specified KIE container. If the solver is still running, this request terminates the solver and then disposes it.
+
+.Request parameters
+[cols="25%,45%,15%,15%", frame="all", options="header"]
+|===
+|Name
+|Description
+|Type
+|Requirement
+
+|`containerId`
+|ID of the KIE container where the solver resides
+|String
+|Required
+
+|`solverId`
+|ID of the solver
+|String
+|Required
+|===
+
+*<@Jiri: The request works but the server responds "undocumented". This example JSON response body below is what the server needs to say.>*
+
+.Example server response (JSON)
+[source,json]
+----
+{
+  "response": [
+    {
+      "type": "SUCCESS",
+      "msg": "Solver solver1 successfully disposed."
+    }
+  ]
+}
+----
+
+== [POST] /containers/{containerId}/solvers/{solverId}/state/solving
+
+Starts the specified solver within its KIE container to solve a given planning problem.
+
+.Request parameters
+[cols="25%,45%,15%,15%", frame="all", options="header"]
+|===
+|Name
+|Description
+|Type
+|Requirement
+
+|`containerId`
+|ID of the KIE container where the solver resides
+|String
+|Required
+
+|`solverId`
+|ID of the solver
+|String
+|Required
+
+|*body*
+|A marshalled planning problem, such as `EmployeeRoster`, to be optimized by the solver
+|Request body
+|Required
+|===
+
+NOTE: To get the best solution for the planning problem, you can first send a `GET` request to `\http://SERVER:PORT/kie-server/services/rest/server/containers/{containerId}/solvers/{solverId}/bestsolution` and then run this `POST` request accordingly.
+
+*<@Jiri: Validate the JSON request body and provide example JSON response.>*
+
+.Example request body (JSON)
+[source,json]
+----
+{
+  "employeerostering.employeerostering.EmployeeRoster": {
+    "employeeList": {
+      "employeerostering.employeerostering.Employee": [
```

Comment:
This is a mistake in the format. The value of the `employeeList` attribute should be an array, currently it's an object with `employeerostering.employeerostering.Employee` field whose value is the expected array. So there's an extra nesting level. Corrected version:
```json
{
  "employeerostering.employeerostering.EmployeeRoster": {
    "employeeList": [
      {
        "name": "John",
        "skills": {
          "employeerostering.employeerostering.Skill": {
            "name": "reading"
          }
        }
      },
      {
        "name": "Mary",
        "skills": {
          "employeerostering.employeerostering.Skill": {
            "name": "writing"
          }
        }
      },
      {
        "name": "Petr",
        "skills": {
          "employeerostering.employeerostering.Skill": {
            "name": "speaking"
          }
        }
      }
    ],
    "shiftList": [
      {
        "timeslot": {
          "startTime": "2017-01-01T00:00:00",
          "endTime": "2017-01-01T01:00:00"
        },
        "requiredSkill": {
          "-reference": "../../../employeeList/employeerostering.employeerostering.Employee/skills/employeerostering.employeerostering.Skill"
        }
      },
      {
        "timeslot": {
          "-reference": "../../employeerostering.employeerostering.Shift/timeslot"
        },
        "requiredSkill": {
          "-reference": "../../../employeeList/employeerostering.employeerostering.Employee[3]/skills/employeerostering.employeerostering.Skill"
        }
      },
      {
        "timeslot": {
          "-reference": "../../employeerostering.employeerostering.Shift/timeslot"
        },
        "requiredSkill": {
          "-reference": "../../../employeeList/employeerostering.employeerostering.Employee[2]/skills/employeerostering.employeerostering.Skill"
        }
      }
    ],
    "skillList": {
      "employeerostering.employeerostering.Skill": [
        {
          "-reference": "../../employeeList/employeerostering.employeerostering.Employee/skills/employeerostering.employeerostering.Skill"
        },
        {
          "-reference": "../../employeeList/employeerostering.employeerostering.Employee[3]/skills/employeerostering.employeerostering.Skill"
        },
        {
          "-reference": "../../employeeList/employeerostering.employeerostering.Employee[2]/skills/employeerostering.employeerostering.Skill"
        }
      ]
    },
    "timeslotList": {
      "employeerostering.employeerostering.Timeslot": {
        "-reference": "../../shiftList/employeerostering.employeerostering.Shift/timeslot"
      }
    }
  }
}
```
The same goes for `shiftList` a few lines below. However, after fixing the body format I found out that kie-server fails to deserialize timeslot objects because their type is `java.time.LocalDateTime`. Should be easy to fix. I'll file a Jira for that.

Reply:
https://issues.jboss.org/browse/PLANNER-1356.